gr-digital/lib/crc16_async_bb_impl.cc

```diff
@@ -0,0 +1,105 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2021 Cesar Martinez.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ * 
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "crc16_async_bb_impl.h"
+#include <gnuradio/io_signature.h>
+#include <volk/volk.h>
+
+namespace gr {
+namespace digital {
+
+crc16_async_bb::sptr crc16_async_bb::make(bool check)
+{
+    return gnuradio::make_block_sptr<crc16_async_bb_impl>(check);
+}
+
+crc16_async_bb_impl::crc16_async_bb_impl(bool check)
+    : block("crc16_async_bb", io_signature::make(0, 0, 0), io_signature::make(0, 0, 0)),
+      d_npass(0),
+      d_nfail(0)
+{
+    d_in_port = pmt::mp("in");
+    d_out_port = pmt::mp("out");
+
+    message_port_register_in(d_in_port);
+    message_port_register_out(d_out_port);
+
+    if (check)
+        set_msg_handler(d_in_port, [this](pmt::pmt_t msg) { this->check(msg); });
+    else
+        set_msg_handler(d_in_port, [this](pmt::pmt_t msg) { this->calc(msg); });
+}
+
+crc16_async_bb_impl::~crc16_async_bb_impl() {}
+
+void crc16_async_bb_impl::calc(pmt::pmt_t msg)
+{
+    // extract input pdu
+    pmt::pmt_t meta(pmt::car(msg));
+    pmt::pmt_t bytes(pmt::cdr(msg));
+
+    unsigned int crc;
+    size_t pkt_len(0);
+    const uint8_t* bytes_in = pmt::u8vector_elements(bytes, pkt_len);
+    std::vector<uint8_t> bytes_out(2 + pkt_len);
+
+    d_crc_ccitt_impl.reset();
+    d_crc_ccitt_impl.process_bytes(bytes_in, pkt_len);
+    crc = d_crc_ccitt_impl();
+    memcpy((void*)bytes_out.data(), (const void*)bytes_in, pkt_len);
+    memcpy((void*)(bytes_out.data() + pkt_len),
+           &crc,
+           2); // FIXME big-endian/little-endian, this might be wrong
+
+    pmt::pmt_t output = pmt::init_u8vector(
+        pkt_len + 2,
+        bytes_out.data()); // this copies the values from bytes_out into the u8vector
+    pmt::pmt_t msg_pair = pmt::cons(meta, output);
+    message_port_pub(d_out_port, msg_pair);
+}
+
+void crc16_async_bb_impl::check(pmt::pmt_t msg)
+{
+    // extract input pdu
+    pmt::pmt_t meta(pmt::car(msg));
+    pmt::pmt_t bytes(pmt::cdr(msg));
+
+    unsigned int crc;
+    size_t pkt_len(0);
+    const uint8_t* bytes_in = pmt::u8vector_elements(bytes, pkt_len);
+
+    d_crc_ccitt_impl.reset();
+    d_crc_ccitt_impl.process_bytes(bytes_in, pkt_len - 2);
+    crc = d_crc_ccitt_impl();
+    if (crc != *(unsigned int*)(bytes_in + pkt_len - 2)) { // Drop package
+        d_nfail++;
+        return;
```

Comment:
This is probably a very personal matter of style. For me it is completely OK to provide an early return in case some pre-condition does not hold (like in this case, that the CRC is not valid). A (questionable) advantage of the early return is that it avoids an extra level of indentation for the rest of the code.

Reply:
with all of the "this is like the majority of the codebase" comments, I suspected that might be the case. However, unlike some of the stylistic areas where the choice is purely subjective, cyclomatic complexity is something that can be objectively measured. 

On the whole, lower complexity is better; less complex code is easier to read and easier to test. Your point about the extra layer of indentation is the most common excuse, but it's masking code that is too complex and should be decomposed more. Pascal's [famous quote](https://quoteinvestigator.com/2012/04/28/shorter-letter/) applies here; writing good code is hard, but the effort required should rarely be justification for choosing an objectively less correct approach.

I'm nobody; the project is free to code as it sees fit. I offered my comments in the spirit of education. Accept them or reject them as you see fit.